Add ethernet support to eval-pqmon

romandariana opened the pull request:
## Pull Request Description

This PR adds support for the W5500 Ethernet controller chip, providing standard Ethernet connectivity for no-os projects. The implementation includes:

- Core W5500 driver with SPI interface and socket management
- Network interface adapter to integrate with the no-os network_interface API
- IIO subsystem support for W5500-based network communications
- Integration with the eval-pqmon project, adding a new "ethernet" build option

## PR Type
- [x] New feature (change that adds new functionality)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have complied with the [Submission Checklist](http://analogdevicesinc.github.io/no-OS/contributing.html#submission-checklist)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

buha commented on drivers/net/w5500/w5500.h:
```diff
@@ -0,0 +1,352 @@
+#include "network_interface.h"
+
+#include <errno.h>
+#include <string.h>
```
i think you can move errno.h and string.h in the .c file, possibly also delay.h and alloc.h, to clean things up a bit

buha reviewed (commented):
add a builds.json

i see this project doesn't have one, but i would add it at this stage to ensure we don't break this project or the w5500 networking infrastructure

buha commented:
and make sure to review the #ifdefs you added, they seem to break some existing projects that use networking

romandariana commented on drivers/net/w5500/w5500.c:
```diff
@@ -0,0 +1,1118 @@
+		val2 = (data[0] << 8) | data[1];
+
+		attempts++;
+	} while (val1 != val2 && attempts < max_attempts);
```
The documentation recommends reading it twice. Using a loop seems a bit too much indeed. Should I keep the double reading and return an error if the 2 values are different, without the loop?

Regarding the builds.json, I think it was skipped because an external library is provided at build time. For example:

    "pqmon_ethernet": {
      "flags": "INTERFACE=ethernet TARGET=max32650 BASIC_EXAMPLE=y IIOD=y HEAP_SIZE=0x80000 STACK_SIZE=0x20000 PQLIB_PATH=/path/to/pqlib"
    },
    
How should I handle the pqlib?

buha replied:
1st question, ok, if the datasheet recommends that (very weird!) keep a 2 retry counter then

2nd question, i think we have the infrastructure for this but it has to be discussed with the CI team.

buha replied:
can you give me a page number where that retry is documented ? i searched and couldn't find anything in the datasheet

romandariana replied:
"Therefore, it is recommended that you read all 16-bits twice or more until getting 
the same value."
I found this on page 54  /  67

buha replied:
that's crazy, it even says twice or more :D ok, let's leave it as-is.

romandariana commented:
I pushed a v2 with some fixes for now:

v1 -> v2
 - use /** instead of /*! doxygen syntax in header files
 - moved errno.h and string.h in the .c file, same for delay.h and alloc.h
 - instead of using network_interface.h, i refactored the driver to use driver specific structure for socket address to see if we get rid of some of the build errors this way. a low-level driver would better now depend on higher level network interface. the mapping between the two is done in w5500_network

romandariana commented:
v2 -> v3

 - add #ifdef NO_OS_W5500_NETWORKING to w5500_network files to avoid getting it built in random projects

buha commented:
ok, i'm dropping the requirement to add builds.json at this point, that's a separate discussion because any build of this project requires pqlib and i am not sure we have the infrastructure now, that can be a separate pr

CiprianRegus commented on network/w5500_network/w5500_network.c:
```diff
@@ -0,0 +1,643 @@
+
+	socket = &net_dev->sockets[physical_id];
+
+	socket->role = (proto == PROTOCOL_TCP) ? W5500_ROLE_CLIENT : W5500_ROLE_UNUSED;
```
I don't think you can know if the socket will be used by a client or server at this point

CiprianRegus commented on projects/eval-pqmon/src/examples/basic/basic_example.c:
```diff
@@ -225,6 +225,23 @@ int basic_pqm_firmware()
+	struct w5500_dev *mac_dev;
+	struct w5500_network_dev *net_dev;
+
+	status = w5500_init(&mac_dev, &w5500_ip);
```
I think you can move this part in the `w5500_network_init()` function in order to simplify the application code.

CiprianRegus commented on projects/eval-pqmon/src/common/common_data.c:
```diff
@@ -33,6 +33,47 @@
+	.gpio_reset = w5500_rst_gpio_ip,
+	.gpio_int = w5500_int_gpio_ip,
+	.mac_addr = {0x00, 0x08, 0xDC, 0x01, 0x02, 0x03},
+	.ip_addr = {192, 168, 1, 110},
```
Is there a way to support DHCP besides the static IP assignment? Also, it would be nice if you could define a macro for the IP address and netmask (similarly to this approach: https://github.com/analogdevicesinc/no-OS/blob/main/projects/apard32690/Makefile#L13-L16 , however you can set it directly into the init param). I think that would provide a simple way of building the project with different ip addresses for the server.

romandariana commented on network/w5500_network/w5500_network.c:
```diff
@@ -0,0 +1,643 @@
+	if (ret)
+		return ret;
+
+	new_socket = &net_dev->sockets[new_physical_id];
+	new_socket->sock_id = sock_id;
+	new_socket->local_port = server_socket->local_port;
+	new_socket->in_use = 1;
+	new_socket->role = W5500_ROLE_SERVER;
+
+	server_socket->sock_id = net_dev->next_virtual_id++;
+	server_socket->role = W5500_ROLE_CLIENT;
```
If I understood correctly the iio module where accept is called expects a new client socket to be created and the server socket keeps listening for new connections. But for W5500, the socket that is listening changes status and becomes a "client" socket. To bypass this, in w5500 accept function I created a new socket that will be the server socket, thus the 3 function callings above, then I switch the virtual ids. Should I extract this switch in another static function to make things clearer?

CiprianRegus replied:
I'd say just adding a comment explaining this behavior would be good enough.

CiprianRegus commented on drivers/net/w5500/w5500.c:
```diff
@@ -0,0 +1,1099 @@
+int w5500_reg_read(struct w5500_dev *dev, uint8_t block,
+		   uint16_t addr, uint8_t *data, uint16_t len)
+{
+	uint8_t spi_tx[3];
+	uint8_t spi_rx[3 + len];
```
You can use just one buffer.

CiprianRegus commented on drivers/net/w5500/w5500.c:
```diff
@@ -0,0 +1,1099 @@
+	if (ret)
+		return ret;
+
+	ret = w5500_reg_write(dev, W5500_COMMON_REG, W5500_RCR, &retry_count, 1);
+	if (ret)
+		return ret;
+
+	return 0;
```
You can return directly

CiprianRegus commented on drivers/net/w5500/w5500.c:
```diff
@@ -0,0 +1,1099 @@
+	    status != W5500_Sn_SR_MACRAW)
+		return -ENOTCONN;
+
+	while (total_sent < len) {
```
Would it be possible to remain stuck in this loop (e.g the other side disconnects while the buffer is full)? If so, I'd add a timeout.

CiprianRegus commented on drivers/net/w5500/w5500.c:
```diff
@@ -0,0 +1,1099 @@
+ *
+ * @return Number of bytes received on success, negative error code otherwise
+*******************************************************************************/
+int w5500_socket_recv(struct w5500_dev *dev, uint8_t sock_id, void *buf,
```
By default, `recv()` is a blocking function. I think you should implement it the same way you did with `send()` above.

CiprianRegus commented on drivers/net/w5500/w5500.c:
```diff
@@ -0,0 +1,1099 @@
+ *
+ * @return 0 on success, negative error code otherwise
+*******************************************************************************/
+int w5500_socket_sendto(struct w5500_dev *dev, uint8_t sock_id, const void *buf,
```
This should block as well.

CiprianRegus commented on drivers/net/w5500/w5500.c:
```diff
@@ -0,0 +1,1099 @@
+	memcpy(dev->subnet_mask, init_param->subnet_mask, 4);
+	memcpy(dev->gateway, init_param->gateway, 4);
+
+	memcpy(&dev->retry_time, &init_param->retry_time, 1);
+	memcpy(&dev->retry_count, &init_param->retry_count, 1);
```
The W5500 seems to have some default values for these that look reasonable. I think these 2 fields in the init parameter should be "optional" in the sense that if they are 0, you don't actually write the register. This way you don't force the users to check the datasheet.

CiprianRegus commented on drivers/net/w5500/w5500.c:
```diff
@@ -0,0 +1,1099 @@
+	struct w5500_dev *dev;
+	int ret;
+
+	dev = (struct w5500_dev *)malloc(sizeof(*dev));
```
Use `no_os_calloc()` instead to zero out the memory.

romandariana commented:
v5 -> v6
    - solve the dev leak in w5500_network_init()

romandariana commented:
v6 -> v7:
   - new commit with builds.json

romandariana commented:
v7 -> v8 -> v9:
    - add mac_dev clean up in w5500_network_init()
    
Note: If the maxim projects builds fails again, I am pretty sure it is not because of these commits. I saw another pr builds fail because of the same project error.
